### ISSUE TITLE
 TypeError: only integer tensors of a single element can be converted to an index

### ISSUE THREAD

**Issue by edwinlu-hdu** [Schindler-EPFL-Lab/thermo-nerf#5]
Description:
I encountered the following error while running the program:
density = self.average_init_density * trunc_exp(density_before_activation.to(positions))
TypeError: only integer tensors of a single element can be converted to an index

It seems that there might be an issue with converting a tensor to an index. Could you please provide some insights or guidance on how to resolve this issue? Any assistance would be greatly appreciated.
Thank you for your attention to this matter.

**Comment by YunSeok-Kang** [comment]
I have identified and fixed the issue causing the TypeError in the latest pull request. 

**Root Cause:**
The error was due to `self.average_init_density` being set to a string value 'tcnn' instead of a numeric value. This caused a TypeError when attempting to use it in calculations.

**Solution:**
The `ThermalNerfactoTField` class initialization has been updated to correctly pass a numeric value for `average_init_density`. Specifically, the `super().__init__` call in `ThermalNerfactoTField` now includes the correct numeric parameter.

Here is the relevant change:
```python
super().__init__(
    aabb,
    num_images,
    num_layers,
    hidden_dim,
    geo_feat_dim,
    num_levels,
    base_res,
    max_res,
    log2_hashmap_size,
    num_layers_color,
    num_layers_transient,
    features_per_level,
    hidden_dim_color,
    hidden_dim_transient,
    appearance_embedding_dim,
    transient_embedding_dim,
    use_transient_embedding,
    use_semantics,
    num_semantic_classes,
    pass_semantic_gradients,
    use_pred_normals,
    use_average_appearance_embedding,
    spatial_distortion,
    1.0,  # Correct numeric value for average_init_density
    implementation,
)

**Comment by Neal2020GitHub** [comment]
@YunSeok-Kang 

Hi, it works for the bug. However, it will raise the following error after some steps. Have you encountered this? Thank you. 

```
mlflow.exceptions.MlflowException: Got invalid value 21.10323715209961 for metric 'psnr' (timestamp=1722956314548). Please specify value as a valid double (64-bit floating point)
```

**Comment by MalcolmMielle** [comment]
Sorry for the delayed answer. At my lab, we will in the following week reproduce this work. I'll let you know if we encounter this error. In any case, don't hesitate to push a PR if you find the solution before that

**Comment by YunSeok-Kang** [comment]
Hi @Neal2020GitHub,

I also encountered the same issue and was able to resolve it using the method I'll describe below. I'm not entirely sure if it's the best solution, but it worked for me, so I thought I'd share it with you.

The error you're seeing seems to be related to how `mlflow` handles metric values, particularly when dealing with tensors in PyTorch. I made a modification to the `_validate_metric` function in the `mlflow/utils/validation.py` file. The updated code includes a check to see if the `value` is a PyTorch tensor, and if so, it converts the tensor to a float using the `.item()` method. Here's the code I used:

```python
def _validate_metric(key, value, timestamp, step):
    """
    Check that a metric with the specified key, value, timestamp, and step is valid and raise an
    exception if it isn't.
    """
    _validate_metric_name(key)
    # value must be a Number
    # since bool is an instance of Number check for bool additionally
    
    if type(value).__module__ == 'torch' and type(value).__name__ == 'Tensor':
        value = value.item()  # Convert to float if it's a tensor
    
    if not _is_numeric(value):
        raise MlflowException(
            f"Got invalid value {value} for metric '{key}' (timestamp={timestamp}). "
            "Please specify value as a valid double (64-bit floating point)",
            INVALID_PARAMETER_VALUE,
        )

    if not isinstance(timestamp, numbers.Number) or timestamp < 0:
        raise MlflowException(
            f"Got invalid timestamp {timestamp} for metric '{key}' (value={value}). "
            "Timestamp must be a nonnegative long (64-bit integer) ",
            INVALID_PARAMETER_VALUE,
        )

    if not isinstance(step, numbers.Number):
        raise MlflowException(
            f"Got invalid step {step} for metric '{key}' (value={value}). "
            "Step must be a valid long (64-bit integer).",
            INVALID_PARAMETER_VALUE,
        )
```
This change helped me resolve the issue by ensuring that the value is correctly processed as a float, which prevented the MlflowException you mentioned. I hope this solution works for you as well, but please let me know if you find a better approach or if you encounter any further issues!

**Comment by YunSeok-Kang** [comment]
Hi @MalcolmMielle,

Thank you for your response. I've already addressed the other issue and submitted a new Pull Request. However, for the issue raised by @Neal2020GitHub, I had to modify the internal code of the external package `mlflow` to resolve the problem. Because of this, I couldn't commit those changes to the repository. I believe there might be a better way to solve this issue without modifying the package code directly.

Looking forward to your thoughts on this!

**Comment by YunSeok-Kang** [comment]
I just realized that @Neal2020GitHub had already opened Issue #7, and I mistakenly responded here without noticing it. While I'm not sure if the solution I provided is the best one, I was able to resolve the issue using the method I described in my comment on Issue #7. If you find a better solution, please feel free to share it with me!

**Comment by MalcolmMielle** [comment]
Fixed by #8
